src/arch_cuda.jl

```diff
@@ -0,0 +1,108 @@
+using CUDA, CUDA.CUSPARSE, CUDA.CUFFT
+import ..Architectures as AC
+import ..BoundaryConditions as BC
+import ..Solvers as SO
+import ..Utils as UT
+import SparseArrays: SparseMatrixCSC
+import KernelAbstractions: __iterspace, __groupindex, __dynamic_checkbounds,
+                           __validindex, CompilerMetadata
+import ..DistributedComputations: Distributed
+
+const GPUVar = Union{CuArray, CuContext, CuPtr, Ptr}
+
+
+const CUDAGPU = AC.GPU{<:CUDABackend}
+
+CUDAGPU() = GPU(CUDABackend(always_inline=true))
+
+# Keep default CUDA backend
+function AC.GPU()
+    if CUDA.has_cuda_gpu()
+        return CUDAGPU()
+    else
+        msg = """We cannot make a GPU with the CUDA backend:
+                 a CUDA GPU was not found!"""
+        throw(ArgumentError(msg))
+    end
+end
+
+
+Base.summary(::CUDAGPU) = "CUDAGPU"
+
+AC.architecture(::CuArray) = CUDAGPU()
+AC.architecture(::CuSparseMatrixCSC) = AC.GPU()
+AC.architecture(::SparseMatrixCSC)   = AC.CPU()
+AC.array_type(::AC.GPU{CUDABackend}) = CuArray
+
+AC.on_architecture(::AC.CPU, a::CuArray) = Array(a)
+
+AC.on_architecture(::CUDAGPU, a::Array) = CuArray(a)
+AC.on_architecture(::CUDAGPU, a::CuArray) = a
+AC.on_architecture(::CUDAGPU, a::BitArray) = CuArray(a)
+AC.on_architecture(::CUDAGPU, a::SubArray{<:Any, <:Any, <:CuArray}) = a
+AC.on_architecture(::CUDAGPU, a::SubArray{<:Any, <:Any, <:Array}) = CuArray(a)
+AC.on_architecture(::AC.CPU, a::SubArray{<:Any, <:Any, <:CuArray}) = Array(a)
+AC.on_architecture(::CUDAGPU, a::StepRangeLen) = a
```

Comment:
In the long run this ought to be `adapt(backend, a)`

Reply:
yeah that would be a big negative diff to combine `on_architecture` and `adapt`